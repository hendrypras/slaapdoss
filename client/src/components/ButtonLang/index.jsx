import PropTypes from 'prop-types';
import classNames from 'classnames';
import { KeyboardArrowRightOutlined as KeyboardArrowRightOutlinedIcon, CheckOutlined } from '@mui/icons-material';
import { Avatar, Modal, Fade, MenuItem, Divider, ListSubheader } from '@mui/material';
import { FormattedMessage } from 'react-intl';
import { useState } from 'react';
import { useDispatch } from 'react-redux';

import { setLocale } from '@containers/App/actions';

import classes from './style.module.scss';

const ButtonLang = ({ locale, className }) => {
  const dispatch = useDispatch();
  const [openModal, setOpenModal] = useState(false);

  const handleOpenModal = () => {
    setOpenModal(true);
  };

  const handleCloseModal = () => {
    setOpenModal(false);
  };

  const onSelectLang = (lang) => {
    if (lang !== locale) {
      dispatch(setLocale(lang));
    }
    handleCloseModal();
  };

  return (
    <>
      <button
        data-testid="button-lang"
        type="button"
        className={classNames(classes.toggle, className)}
        onClick={handleOpenModal}
      >
        <Avatar className={classes.avatar} src={locale === 'id' ? '/id.png' : '/en.png'} />
        <div className={classes.lang}>{locale}</div>
        <KeyboardArrowRightOutlinedIcon className={classes.arrowIcon} />
      </button>
      <Modal
        open={openModal}
        onClose={handleCloseModal}
        slotProps={{
          backdrop: {
            sx: {
              backdropFilter: 'blur(3px)',
            },
          },
        }}
        className={classes.modal}
      >
        <Fade in={openModal}>
          <div className={classes.paper}>
            <ListSubheader className={classes.headTitle}>Pilih Bahasa</ListSubheader>
            <Divider />
            <MenuItem onClick={() => onSelectLang('id')} selected={locale === 'id'} className={classes.menu}>
              <div className={classes.wrapperLeft}>
                <Avatar className={classes.avatar} src="/id.png" />
                <div className={classes.textLang}>
                  <FormattedMessage id="app_lang_id" />
                </div>
              </div>
              {locale === 'id' && <CheckOutlined className={classes.checkIcon} />}
            </MenuItem>
            <Divider />
            <MenuItem onClick={() => onSelectLang('en')} selected={locale === 'en'} className={classes.menu}>
              <div className={classes.wrapperLeft}>
                <Avatar className={classes.avatar} src="/en.png" />
                <div className={classes.textLang}>
                  <FormattedMessage id="app_lang_en" />
                </div>
              </div>
              {locale === 'en' && <CheckOutlined className={classes.checkIcon} />}
            </MenuItem>
            <Divider />
          </div>
        </Fade>
      </Modal>
    </>
  );
};

ButtonLang.propTypes = {
  locale: PropTypes.string.isRequired,
  className: PropTypes.string,
};

export default ButtonLang;
