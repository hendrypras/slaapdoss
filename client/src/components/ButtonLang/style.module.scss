@import '@styles/base';

.toggle {
  @include displayFlex(row, center, center);
  padding: 0.25rem;
  border: none;
  border-radius: 1rem;
  background-color: $color-white;
  cursor: pointer;
  color: $color-black-100;
  gap: 0.2rem;
  box-shadow: rgba(0, 0, 0, 0.3) 0px 2px 8px 0px;
  .avatar {
    width: 1.5rem;
    height: 1.5rem;
  }
  .lang {
    font-family: Nunito, sans-serif;
    font-weight: 900;
    font-size: 0.85rem;
    text-transform: uppercase;
    color: $color-black-100;
  }
  .arrowIcon {
    font-size: 1.3rem;
    color: $color-black-100;
  }
}
.modal {
  display: flex;
  align-items: center;
  justify-content: center;
}

.paper {
  background-color: white;
  box-shadow:
    24px 38px 48px 0 rgba(0, 0, 0, 0.14),
    0 9px 46px 8px rgba(0, 0, 0, 0.12),
    11px -15px 66px 0 rgba(0, 0, 0, 0.2);
  padding: 0 1.2rem 0.7rem 1.2rem;
  width: 400px;
  overflow: auto;
  border-radius: 1rem;
  max-width: 80%;
  @include screen(xl) {
    max-width: 90%;
  }
  .headTitle {
    font-size: 1.2rem;
    font-weight: 600;
    padding: 0.4rem 1rem;
    color: $color-black-100;
  }
  .menu {
    @include displayFlex(row, space-between, center);
    padding: 1rem;
    width: 100%;
    gap: 0.2rem;
    .wrapperLeft {
      @include displayFlex(row, start, center);
      gap: 1rem;
      .avatar {
        width: 1.5rem;
        height: 1.5rem;
      }
      .textLang {
        font-size: 1rem;
        font-weight: 600;
        color: $color-black-100;
      }
    }
    .checkIcon {
      color: $color-green-100;
    }
  }
}
