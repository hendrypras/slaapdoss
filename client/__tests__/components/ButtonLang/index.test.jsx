import { render } from '@utils/testHelper';
import ButtonLang from '@components/ButtonLang';

let wrapper;
const mockProps = {
  className: 'class',
  locale: 'id',
};

beforeEach(() => {
  wrapper = render(<ButtonLang className={mockProps.className} locale={mockProps.locale} />);
});

describe('Button', () => {
  test('Should render correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('button-lang')).toBeInTheDocument();
  });
  test('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
