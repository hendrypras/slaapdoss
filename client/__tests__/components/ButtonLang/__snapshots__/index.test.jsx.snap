// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="toggle class"
        data-testid="button-lang"
        type="button"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular avatar css-1wlk0hk-MuiAvatar-root"
        >
          <img
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="/id.png"
          />
        </div>
        <div
          class="lang"
        >
          id
        </div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium arrowIcon css-i4bv87-MuiSvgIcon-root"
          data-testid="KeyboardArrowRightOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="toggle class"
      data-testid="button-lang"
      type="button"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular avatar css-1wlk0hk-MuiAvatar-root"
      >
        <img
          class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
          src="/id.png"
        />
      </div>
      <div
        class="lang"
      >
        id
      </div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium arrowIcon css-i4bv87-MuiSvgIcon-root"
        data-testid="KeyboardArrowRightOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
